core/src/main/scala/org/apache/spark/internal/config/package.scala

```diff
@@ -2429,4 +2429,21 @@ package object config {
       .version("3.4.0")
       .timeConf(TimeUnit.MILLISECONDS)
       .createWithDefaultString("5s")
+
+  private[spark] val CUSTOM_SERVICE_PORT =
+    ConfigBuilder("spark.service.port.custom")
+      .doc("Customize the spark service port range, " +
+        "service like SparkUI.port, SparkDriver.port, NettyService.port etc." +
+        "The starting port number is spark.service.port.custom.origin " +
+        "and the port range is limited by spark.port.maxRetries. " +
+        "e.g. spark.service.port.custom.origin=49152 and spark.port.maxRetries=10, " +
+        "the actual range of the port is greater than or equal to 49152 and less than 49163")
+      .booleanConf
+      .createWithDefault(false)
+
+  private[spark] val CUSTOM_SERVICE_PORT_ORIGIN =
+    ConfigBuilder("spark.service.port.custom.origin")
+      .intConf
+      .checkValue(v => v >= 1024, "The threshold should be greater than or equal to 1024.")
+      .createWithDefault(49152)
```

Comment:
Should probably have one option that's defined with `createOptional`.

BTW, why do we need this though? can't we simply use 0 to automatically find the availalbe port?

Reply:
Thanks for your review, the default value of 41952 is setted because there are some ideas that default to 1024-> 49151 as Registered Ports and 49152-> 65535 as Dynamic and/or Private Ports, Of course this default value needs to be discussed again.

The port number of 0, which we know is automatically assigned by the kernel, is not applicable in the following scenario: the yarn cluster and the spark client are two separate clusters, and for security reasons, the firewall restricts port access. An exact port value is required and we cannot provide it to the relevant ops staff after the port has been randomly confirmed, so the port needs to be restricted to a certain range in advance for retries.

Reply:
You can just make it optional instead of deciding a default value here (e.g., if it's set, enable this feature).

If you need to specify a specific range, the configuration would have to be able to specify both lower bound and upper bound.

Reply:
About spark.service.port.custom.origin default value, I guess I have your ideas, It will look more elegant.
In addition, upper bound has been limited, the final port range is: 
spark.service.port.custom.origin <= port.range <= spark.service.port.custom.origin + spark.port.maxRetries